rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,401 @@
+for the DP Browser App to facilitate the uploading of files. It will also enqueue shared links for downloading.
+This additional functionality can be built into the existing DP Backend Lambda. 
+
+The backend will also be responsible for performing validation of the file being uploaded. The validation is
+performed during the initial request to upload. 
+
+#### Download Queue
+
+The Download Queue is an [AWS SQS](https://aws.amazon.com/sqs/) that will track pending and in-progress download jobs.
+The download jobs are used by the EC2 instances to know what files to download before uploading them to S3.
+
+##### Download Job Entry
+
+These are the fields that will be in the download job placed in the queue:
+
+- Link - a to the link to download. Use to retrieve the file from CSP.
+- submission_id - identifies the submission. Used to determine the storage location.
+- dataset_id - identifies the dataset. Used to determine the storage location
+- file_name - the name of the file being downloaded
+
+#### AWS STS
+
+The AWS STS service will be used to generate temporary credentials used to upload a file to S3. The backend
+server will use the STS [assume role](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html)
+functionality to generate a token with temporary access to upload to S3. This token is referred to as an access token.
+
+The access token generated and provided grants direct access to the S3 bucket for uploading. This will allow
+the browser APP to use the AWS JS SDK to perform the multipart uploads to S3 for large files. A backend server will be used
+to generate the token and additional sessions policies will be applied to the token to further restrict access
+
+According to AWS:
+
+> The resulting session's permissions are the intersection of the role's identity-based policies and the session policies.
+
+The IAM policies bellow will be combined to limit the access of the token generated to only allow the bearer to upload the
+file in the session policy resource. The session policy will be generated at the time of the request and will use the
+asset_uuid generated by the server and the data set_uuid provided in the request.
+
+##### Assumed Role IAM Policy
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:ListMultipartUploadParts",
+        "s3:PutObject",
+        "s3:AbortMultipartUpload"
+      ],
+      "Resource": ["arn:aws:s3:::upload-bucket/"]
+    }
+  ]
+}
+```
+
+##### IAM Session Policy
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:ListMultipartUploadParts",
+        "s3:PutObject",
+        "s3:AbortMultipartUpload"
+      ],
+      "Resource": [
+        "arn:aws:s3:::upload-bucket/{submission_id}/{dataset_id}/{file_name}"
+      ]
+    }
+  ]
+}
+```
+
+#### EC2
```

Comment:
Will the EC2 instance be spun up for this tasks, or always up waiting for a download request?
In the rare case where multiple submission are happening at the same time, will they be done one at a time by a single EC2 instance, or will there be multiple EC2 instances working simultaneously.   We may want to design in a way to auto scale if needed.

Reply:
Ah nevermind, you address this later.